### ISSUE TITLE
Building Zig + C++ mixed projects fail to compile on macOS

### ISSUE THREAD

**Issue by prime31** [ziglang/zig#6812]
This is related to the regression from #6773. C++ compilation fails with just `Segmentation fault: 11`. As mentioned in #6773, `0.6.0+68bf29c31` is the last working version I have that can compile C++ code on macos.

You can reproduce by just testing on this repo (https://github.com/prime31/zig-upaya) that has C code (Sokol) and C++ (Dear ImGui) mixed with the zig. I made a branch that is updated to be compatible with Zig master as of todays Zig version. Repro steps:

```
git clone --recursive https://github.com/prime31/zig-upaya
cd zig-upaya
git checkout zig-master-10-24
zig build
```

**Comment by jedisct1** [comment]
Using the system linker instead of LLVM fixes this.

You can try the [changes for Big Sur](https://github.com/ziglang/zig/issues/6318) even if you are not running Big Sur.

**Comment by prime31** [comment]
@jedisct1 I just built zig using your fork and on Catalina I get the same error: `Segmentation fault: 11`

**Comment by prime31** [comment]
Update: thank you @jedisct1! Your fork and `bigsur` branch fixes the issue!

**Comment by prime31** [comment]
Do we have a PR open for these changes? Is there any reason they aren't already in Zig proper?

**Comment by kubkon** [comment]
A quick update that as part of my effort to bring Zig to Apple Silicon, I'll be incorporating @jedisct1's BigSur patches into the libstd. The reason why I'm hesitant to include the patch as-is is that it includes a call to `xcrun` which shouldn't be done inside the linker itself, but rather in a place like `lib/std/zig/system.zig`.

**Comment by kubkon** [comment]
I've decided to dig a bit deeper, and I believe this is an upstream issue which explains why @jedisct1's BigSur patch works. I've submitted an [upstream bug report](https://bugs.llvm.org/show_bug.cgi?id=48019), and I'll try to check if the upstream `lld` has got it fixed already, and perhaps look for a patch if there's any. If not, I might try and figure out a fix for it if there's none upstream.

In the meantime, however, I'd like to hear you out what to do to patch it in Zig in the meantime so that we don't block the 0.7.0 release.

cc @andrewrk
